docs/proposals/acting-as.md

```diff
@@ -0,0 +1,219 @@
+creates a pod.
+
+<img src="https://www.lucidchart.com/publicSegments/view/14d7b49f-469c-4bc4-8bf7-f13581ee0220/image.png"
+alt="UML Sequence Diagram for Current Replication Controller">
+<!-- Original document: https://www.lucidchart.com/documents/edit/34454705-cff6-4b42-a95f-24d69a36184c
+     contact etune@google.com to request access -->
+
+This diagram shows how it woud work today if a user created a program that was itself a Kubernetes
+client, and acted as an RC, but for a fixed template, number of replicas, etc.
+
+<img src="https://www.lucidchart.com/publicSegments/view/a719bd66-191f-478f-8c42-bc033d19ef25/image.png"
+alt="UML Sequence Diagram for User-written Replication Controller">
+<!-- Original document: https://www.lucidchart.com/documents/edit/92c9f481-a104-4283-8c8d-6bec71a1d9d7
+     contact etune@google.com to request access -->
+
+This diagram shows a proposal for how I think RC-manager should work.  It uses a new header (or
+maybe query parameter?) which tells the apiserver that the RC-manager wants to Act-As
+a different user. It is much closer to the "user-written replication manager" case.
+
+<img src="https://www.lucidchart.com/publicSegments/view/c59b1964-2345-4910-8580-d4c636f28fa3/image.png"
+alt="UML Sequence Diagram for Proposed RC-manager">
+<!-- Original document: https://www.lucidchart.com/documents/edit/7c8cdcde-9803-4211-8cf1-a05c6da3fbc6
+     contact etune@google.com to request access -->
+
+Things to note in the new flow:
+
+- controller-manager is using authorized port instead of the localhost port to talk to 
+   kube-apiserver
+- controller-manager sets a header to say it wants to lower its root-like powers in order
+  to act as less powerful user (which solves the confused deputy).
+
+
+## Argument for proposal
+
+This change would protect us from the confused-deputy problem.  This is important because:
+
+1. we have already had confused deputy bugs in many of our controllers.  there will be others.
+1. this is a common class of bugs in systems such as this.
+1. we might trust ourselves to keep controllers in the main repo free from these bugs;
+   but we will not have time to closely review contributed code for these bugs.
+1. out authorization rules will get more complex over time, so deputies will have
+   to have new, more complex checks in them.  Bug fixing contrib code will be hard,
+   since authorization is compiled into the controller.
+1. Writing a controller in a language other than go difficult because you
+   will not be able to reuse on-controller-auth code.
+
+## Work items
+
+The specific changes needed for this new flow are:
+
+- change controller manager to use the secure port rather than 127.0.0.1 on the
+  master This does require users to change their flags, but we can support the
+  old way for several release cycles.
+- change apiserver to recognize an "Acting-As" header in requests, and to check whether the
+  originally authenticated user is able to act as that user.
```

Comment:
I envisioned only being able to act as a user.  I envision that the steps are:

1. authenticate user as we do today
1. authorize user to act-as another user
1. lookup group membership of (effective) user
1. authorize REST action using effective user and its group membership

I was thinking that we would only let you act-as a kubernetes ServiceAccount, not as another type of account.  Since ServiceAccount is a kubernetes object, it can be referenced in Openshift (or other) Policy statements.  So, you can have a Policy and Role like this:

```
kind: Role
...
rules:
  - resourceNames: [ "namespace/foo/serviceAccounts/default", ...]
  - verb: ["actAs"]
-----
kind: roleBinding
...
usernames: ["system:replicationController"]
```
which says that user "system:replicationController" can "actAs" (or maybe "use") the default service account for namespace foo (and probably every other namespace, as they get added.)